### DISCUSSION TITLE
Matching Multiple Loop Variables

### DISCUSSION THREAD

**Discussion by Abhiram98** [ast-grep/ast-grep#717] [Q&A] [answered]
I would like to match loop iterators in into separate meta variables.

For example, in the below code, I would like to match both 'i' and 'j' into separate meta variables.

```
for i,j in array:
    print(i+j)
```

[Here](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJFgsICRZPTEsMiIsInJld3JpdGUiOiIiLCJjb25maWciOiJydWxlOlxuICBraW5kOiBmb3Jfc3RhdGVtZW50XG4gIGFsbDpcbiAgICAtIGhhczpcbiAgICAgICAgZmllbGQ6IGxlZnRcbiAgICAgICAgcGF0dGVybjogJEksJEpcbiAgICAtIGhhczpcbiAgICAgICAgZmllbGQ6IGJvZHlcbiAgICAgICAgcGF0dGVybjogcHJpbnQoJEkrJEopIiwic291cmNlIjoiZm9yIGksaiBpbiBhcnI6XG4gICAgcHJpbnQoaStqKSJ9) is my unsuccessful attempt in the playground.

Thanks in advance for your help!

Edit: Seems like the [pattern syntax](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicHl0aG9uIiwicXVlcnkiOiJmb3IgJEksICRKIGluICRBUlI6XG4gICAgcHJpbnQoJEkgKyAkSikiLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoicnVsZTpcbiAga2luZDogZm9yX3N0YXRlbWVudFxuICBhbGw6XG4gICAgLSBoYXM6XG4gICAgICAgIGZpZWxkOiBsZWZ0XG4gICAgICAgIHBhdHRlcm46ICRJLCRKXG4gICAgLSBoYXM6XG4gICAgICAgIGZpZWxkOiBib2R5XG4gICAgICAgIHBhdHRlcm46IHByaW50KCRJKyRKKSIsInNvdXJjZSI6ImZvciBpLGogaW4gYXJyOlxuICAgIHByaW50KGkraikifQ==) works but the template syntax doesn't?

**Comment by HerringtonDarkholme** [comment]
I'm not too sure about your question but looks like this can work?

<img width="977" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/c567be92-c65c-4522-8558-06c9e46b663f">

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicHl0aG9uIiwicXVlcnkiOiJmb3IgJEksICRKIGluICRBUlI6ICQkJCIsInJld3JpdGUiOiIiLCJjb25maWciOiJydWxlOlxuICBraW5kOiBmb3Jfc3RhdGVtZW50XG4gIGFsbDpcbiAgICAtIGhhczpcbiAgICAgICAgZmllbGQ6IGxlZnRcbiAgICAgICAgcGF0dGVybjogJEksJEpcbiAgICAtIGhhczpcbiAgICAgICAgZmllbGQ6IGJvZHlcbiAgICAgICAgcGF0dGVybjogcHJpbnQoJEkrJEopIiwic291cmNlIjoiZm9yIGksaiBpbiBhcnI6XG4gICAgcHJpbnQoaStqKSJ9

In YAML rule,

<img width="380" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/0db10d41-8843-465b-8af5-8566c6a3508d">




https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiZm9yICRJLCAkSiBpbiAkQVJSOiAkJCQiLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoicnVsZTpcbiAgcGF0dGVybjogJ2ZvciAkSSwgJEogaW4gJEFSUjogJCQkJyIsInNvdXJjZSI6ImZvciBpLGogaW4gYXJyOlxuICAgIHByaW50KGkraikifQ==

**Comment by Abhiram98** [comment]
Thanks! 

Is there a way I could use this kind of syntax? I'm looking to build a more complex rule on top of this
<img width="387" alt="Screenshot 2023-11-15 at 7 37 00 PM" src="https://github.com/ast-grep/ast-grep/assets/31768552/4ee5aec1-242e-4458-a76e-f977f5027043">

**Reply by HerringtonDarkholme** [reply]
I think the issue is in the pattern part.
$I, $J is not parsed as pattern list. You can try `for $I, $J in []: $$$` and selector.

**Reply by HerringtonDarkholme** [reply, marked as answer]
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiICAgICAgICAgIHByaW50KCRJKyRKKSIsInJld3JpdGUiOiIiLCJjb25maWciOiJydWxlOlxuICBraW5kOiBmb3Jfc3RhdGVtZW50XG4gIGFsbDpcbiAgICAtIGhhczpcbiAgICAgICAgZmllbGQ6IGxlZnRcbiAgICAgICAgcGF0dGVybjogXG4gICAgICAgICAgY29udGV4dDogJ2ZvciAkSSwkSiBpbiBbXTogJCQkJ1xuICAgICAgICAgIHNlbGVjdG9yOiBwYXR0ZXJuX2xpc3RcbiAgICAtIGhhczpcbiAgICAgICAgZmllbGQ6IGJvZHlcbiAgICAgICAgaGFzOlxuICAgICAgICAgIHBhdHRlcm46IHByaW50KCRJKyRKKVxuICAgICAgICAgIHN0b3BCeTogZW5kIiwic291cmNlIjoiZm9yIGksaiBpbiBhcnI6XG4gICAgcHJpbnQoaStqKSJ9

this should work

**Reply by Abhiram98** [reply]
Thanks! That's exactly what I needed
